linode/manager
# feat: [M3-6888] - Add AGLB "Edit Service Target" drawer

## Description 📝
Add ability to edit a Service Target via a drawer. 

## Changes  🔄
- Refactors `CreateServiceTargetDrawer.tsx` to `ServiceTargetDrawer.tsx` and is used in both add and edit modes
- Adds 'Edit' action menu item functionality to open the Service Target Drawer
- Adds integration test coverage for edit functionality
- Updates `apiv4` and `validation` packages in order to validate and submit the PUT request
- Updates mock endpoint for `rest.get('*/v4beta/aglb/:id/service-targets')` to test the edit flow more accurately 

## Preview 📷
![Screenshot 2023-10-23 at 9 37 17 AM](https://github.com/linode/manager/assets/114685994/045416e9-86a0-4bab-bbd0-991aeb492d3e)

### Video 🎥 

https://github.com/linode/manager/assets/114685994/3ce21c30-0ec2-4bc8-b090-1bd1886fbf0a

## How to test 🧪

### Prerequisites
(How to setup test environment)
- Check out this PR.
- Ensure AGLB feature flag is on.
- Turn the MSW on

### Verification steps 
(How to verify changes)
- Go to http://localhost:3000/loadbalancers/0/service-targets.
- Open up the developer console, go to the network tab, and filter by search term "service-targets" for convenience.
- Click on the `...` at the end of the first service target to open the action menu. Select the Edit action. 
- Observe that a drawer opens and fields are populated with data for that service target.
- Edit some fields (update the label, remove a service target, add/remove endpoints) and submit the form.
- Validate that the PUT network request in the developer console contains the updated data. 
   - Due to mocks, we won't see the table update.
- Validate error handling:
   - Attempt to submit a service target without a label. Validate that a validation error appears on the Label field. 
- Validate that clicking Create Service Target still opens the Service Target Drawer, but with no data populated.
- Confirm that the integration test passes for both create and edit service target flows:
```
 yarn cy:run -s "cypress/e2e/core/loadBalancers/load-balancer-service-targets.spec.ts" 
```

## As an Author I have considered 🤔

*Check all that apply*
